회원 가입 관련 기능 구현

sgo8308 opened the pull request:
### **개요**

회원 가입 관련 기능 구현

### **이슈 링크**

#15 

### **작업 사항 (변경 사항)**
- 회원 가입 관련 controller, service, repository 구현
- 회원 가입시 인풋 데이터 Validation 로직 구현
- 아이디, 이메일, 핸드폰 번호 중복 체크 관련 로직 구현

### 테스트 내역 (o/x)

### **리뷰 받고 싶은 내용**

1. 두 명이 같은 아이디로 동시에 가입하는 부분은 DB가 터뜨리고 스프링이 감싼 DuplicateKeyException 예외를 잡아서 처리하도록 구현하였습니다. 다른 방법보다 간단하고 성능적으로도 좋다고 판단했습니다. 다만 아이디, 이메일, 핸드폰 번호가 모두 같은 예외를 터뜨려서 예외 메세지를 통해 구분을 해주었는데 DB와 너무 긴밀하게 엮이는 것 같아서 고민입니다. DB에서 보내는 예외 메세지를 조금만 수정해도 로직이 망가집니다.

2. 클라이언트 쪽에서 타입에 맞지 않는 input 데이터를 보내면 HttpMessageNotReadableException이 터지고 이 예외를 잡아서 문제가 생긴게 어떤 field인지를 응답으로 내리도록 구현했습니다.이 방식은 실제로 클라이언트가 잘못 입력한 값이 무엇인지 모르고, 또 타입이 맞지 않는 값이 여러개일 경우에도 한 개의 값 밖에 응답으로 보내는 문제가 있습니다. 위 문제가 없도록 구현하는게 까다롭고 실제로 클라이언트 측에서 여러 인풋값을 동시에 잘못 보내는 경우는 드물다고 생각되어 일단 이렇게 구현했는데 더 좋은 방법이 있는지 잘 모르겠습니다.

3. 예상하지 못한 예외는 모두 unExceptedExceptionHandler에서 잡아서 INTERNAL_SERVER_ERROR를 내리는데요. 이렇게 하니 스프링에서 기본적으로 여러가지 상황에 대비해서 적절한 예외가 나가도록 한 부분이 없어져서 제가 일일이 처리해야 하는 상황입니다. 예를 들어 이번에 구현하면서 파라미터가 필요한 controller에 파라미터를 넣지 않고 get요청을 보냈더니 MissingServletRequestParameterException이 나왔고 unExceptedExceptionHandler가 잡아서 INTERNAL_SERVER_ERROR를 응답으로 보냈습니다. 사실 BAD_REQUEST가 맞는데 말이죠. 일단은 이 예외도 잡아서 처리하는 로직을 직접 구현해놓긴 했는데 이 예외 말고도 제가 예상하지 못하면서 INTERNAL_SERVER_ERROR가 아닌 예외가 더 있을 수도 있다고 생각되는데요. 보통 실무에서도 이렇게 스프링에 힘을 빌리지 않고 일일이 모든 상황을 대비해서 예외 처리 로직을 짜게되나요?

ps - 먼저 jiwoo_project_setting과 jiwoo_sign_up을 머지했더니 이 풀리퀘스트를 할 때 file changed가 지금까지 작업만 모든 사항이 생겨버렸습니다. 이 풀리퀘스트에 맨 마지막 커밋만 확인하시면 더 리뷰하기 편하실 것 같습니다 !


### 1차 리뷰 후 추가 내용
- GlobalControllerAdvice 관련
ApiSubError는 삭제하고 각 에러마다 클래스를 만들어주었는데요.(InvalidArgumentApiError, InvalidFormatApiError)
이렇게 하니 에러 메시지에 일관성이 좀 깨지는 것 같습니다.
InvalidArgumentApiError은 필드로 validationErrors를 갖고 있고, InvalidFormatApiError은 validationError를 갖고 있습니다. 추가적으로 다른 에러가 생길 때 마다 이 추가 서브 에러들의 이름이 바뀌면 일관성이 없지 않을까요? 기존에는 모두 subApiErrors로 통일했었습니다.
멘토님께서 얘기해주신 하나의 string으로 일관되게 반환할 수 있지만 그럴 경우 InvalidArgumentApiError에서 여러 errorfield와 rejected value가 하나의 문장 안에 들어 있어서 프론트 엔드 쪽에서 파싱하기가 힘들 것 같습니다.

- 중복 체크 관련
중복체크를 /users?loginId= ~ 이런 api로 처리하게끔 구현하고 response로는 id만 나가도록 구현했는데요. 이 /users라는 api가 중복체크 외에도 실제로 여러 유저들을 특정한 조건으로 찾고 싶을 때도 쓸 수 있을 것 같고 여러 용도로 쓰일 수 있을 것 같다는 생각이 들었습니다.
그럴 경우 각 경우마다 response가 달라질 수 있을 것 같다는 생각이 들어 클라이언트 쪽에서 원하는 field만 쿼리 스트링에 담아서 보내는 식으로 구현해야 할까 생각이 들었습니다. response는 그에 따라 동적으로 달라지게 되구요. 구현하게 되면 API는 /users?loginId='abc123'&name='지우'&fields=id,loginId,name,address 이런 식으로 사용될 것 같습니다. 하지만 아직 뚜렷하게 이 API를 필요로 하는 다른 기능이 생각이 나지 않아 오버 엔지니어링일 것 같아 지금과 같이 구현하였습니다. 멘토님은 어떻게 생각하시나요?

sgo8308 committed:
초기 구축 - 프로젝트 폴더 구성

sgo8308 committed:
초기 구축 - application.properties 설정, application-dev.properties 추가

- 공통적인 속성과 profile은 application.properties에서, 환경별 특별한 설정은 개별 파일에서 설정하도록 application.properties 틀을 만듬

- 개발 환경 세팅을 위한 application-dev.properties 파일 추가

sgo8308 committed:
초기 구축 #2 - logback-spring.xml 추가

- 개발 환경, 운영 환경 별로 logback 설정이 다르게 적용되도록 틀 구성
- 기본적인 logback 설정

sgo8308 committed:
초기 구축 #3 - mybatis 연동

application-dev.properties
- mysql연결을 위한 datasource 관련 속성 추가
- mapper 파일 관리하기 편하게 resources/mybatis/mapper아래 모두 모을 수 있도록 설정 추가

build.gradle
- mybatis와 스프링 연동 관련 의존성 해결해주는 mybatis-spring-boot-starter의 최신 버전인 2.2.2버전 추가
- mysql 연결을 위한 mysql용 jdbc driver 추가

Sample, SampleComtroller, SampleRepository, SampleService
- mybatis 연동을 테스트하고 폴더 구조를 나타내기 위한 sample 파일 추가

SampleMapper.xml
- mybatis를 이용한 데이터 저장 테스트용 mapper 추가

sgo8308 committed:
초기 구축 #4 - 구동 환경 나누기

application.properties
- profile local로 변경

application-local.properties, application-prod.properties
- 로컬 환경과, 운영 환경 설정을 위해서 추가

application-dev.properties, application-local.properties
- 기존에 있던 mybatis 연동 관련 설정 application-local.properties 생성 후 이동

logback-spring.xml
- 로컬 환경 관련 logback 설정 추가

sgo8308 committed:
feat: logback 운영 버전 설정에 file appender 추가 (#2)

sgo8308 committed:
feat(logging): logback-spring 파일 분리 (#2)

sgo8308 committed:
feat(logging): colsole-appender 패턴에 날짜 추가 (#2)

sgo8308 committed:
chore(logging): file-appender 패턴 수정(#2)

sgo8308 committed:
refactor(logging): logback-local과 logback-dev는 root를 debug로 변경 (#2)

의도치 않은 예외 및 프레임워크 단의 에러 파악을 위해서

sgo8308 committed:
refactor(logging,resources): file-appender의 하드코딩된 file name과 file path를 변수로 변경

중복되는 파일 이름과 파일 경로를 변수로 변경하고 파일 이름을 명확하게 market-gola-api.log로 변경하였습니다.

sgo8308 committed:
refactor(logging): console-appender의 로그 패턴 변경

로그 메세지가 다음 줄에 출력도는 현상이 있어 다음 줄에 출력이 안되도록 변경하였습니다.

sgo8308 committed:
refactor(logging): 개발 환경에서도 file-appender가 추가되도록 변경

sgo8308 committed:
feat(user, common): 예외 발생시 일관된 예외 형태가 출력될 수 있도록 초기 구성 (#5)

GlobalControllerAdvice.java
- 모든 패키지에서 공통으로 처리할 ExceptionHandler를 위해서만들었습니다.

SampleControllerAdive.java
- 샘플로 현재는 com.flab.marketgola.user 패키지와 그 하위 패키지에서 발생하는 예외를 처리하도록 구성하였습니다.

SampleErrorController.java, SampleErrorResult.java
- 실제로 예외메시지를 잘 보내는지 테스트하기 위해 추가하였습니다.

sgo8308 committed:
Merge pull request #6 from f-lab-edu/jiwoo

폴더 구성, logback 설정, mybatis 연동, 구동 환경 나누기

sgo8308 committed:
Merge branch 'feature/jiwoo_project_setting' into feature/jiwoo_errorpage_setting

sgo8308 committed:
fix(resources): findById()를 통해 Sample 객체를 찾아올 경우 sampleName이 제대로 저장되지 않는 문제 해결

db의 경우 sample_name으로 저장되는데 대응되는 클래스는 sampleName을 갖고 있기 때문에 저장이 제대로
안되었습니다. 따라서 이런 불일치를 해결해주는 설정을 추가해주었습니다.

sgo8308 committed:
feat(mapper, user.repository): SampleRepository에 findById() 추가 (#8)

테스트DB를 테스트하기 위해 findById() 기능을 추가하였습니다.

sgo8308 committed:
feat(test.resources, test.user.repository): 메모리DB 설정, 매 테스트마다 롤백되도록 설정 (#8)

application.properties
- 테스트 전용 application.properties를 만들고 메모리DB를 설정

build.gradle
- 테스트에서도 Lombok이 사용가능하도록 설정을 추가

schma.sql
- 메모리 DB의 초기 테이블을 구성하기 위해 추가

SampleRepositoryTest
- 메모리DB에 mybatis를 통하여 제대로 저장되는지 확인
- 매 테스트마다 롤백되도록 @Transactional을 테스트에 추가

sgo8308 committed:
Merge branch 'feature/jiwoo_project_setting' into feature/jiwoo_testdb_setting

sgo8308 committed:
refactor(resources, test.resources): 가독성을 위해 모든 application.properties를 yaml 파일로 변환

sgo8308 committed:
feat(user, common): BaseException과 FatalException 생성 (#5)

BaseException
* 추후 예외처리를 위해 BaseException 상속하는 모든 예외들은 자신과 관련된 응답 상태코드를 생성시 등록해야 함

FatalException
* Exception을 구분하기 위해 FatalException 생성
* Error level의 log를 남기고 싶은 에외는 FatalException을 상속하면 됨

sgo8308 committed:
feat(common): ErrorResult 변수 추가 (#5)

sgo8308 committed:
feat(common): TimeUtil 추가(#5)

LocalDataTime을 정돈된 형태의 String으로 변환하는 convert()추가

sgo8308 committed:
feat(user, common): 패키지 특화 ControllerAdvice 삭제 후 GlobalControllerAdvice 기능 추가 (#5)

exceptionHandler()
모든 예외를 BaseException을 상속한 예외로 감싸서 던질 시 모든 예외를 처리가능

doLog()
exception의 타입에 따라 다른 log level로 로깅

createErrorResult()
일관된 예외 결과를 만들어주며 예외가 갖고 있는 responseHttpStatus를 응답상태코드로 이용

sgo8308 committed:
feat(user): SampleErrorController mapping 추가, SampleErrorResult 삭제 (#5)

SampleErrorController
FatalException과 GeneralException을 테스트하기 위한 mapping 메소드 추가

SampleErrorResult
삭제 후 ErrorResult 추가

sgo8308 committed:
feat(user): save 후 save된 객체를 받기 위해 mapper와 repository 분리 (#10)

sgo8308 committed:
test(user): 메모리DB 저장 test가 정상 작동하도록 테스트 변경 (#10)

sgo8308 committed:
feat(test.user): 테스트 환경 별 프로파일 분리 (#10)

테스트 클래스에서 직접 원하는 프로파일을 선택할 수 있도록 분리

sgo8308 committed:
test(user): SampleRepositoryTest가 application-unit을 이용하도록 지정 (#10)

sgo8308 committed:
refactor(common, user): BaseException 수정, FatalException 삭제 (#9)

* LogLevel과 HttpStatus를 예외를 생성할 때 유연하게 구성할 수 있도록 변수를 추가 후 생성자 변경
* 바뀐 BaseException에 따라 SampleErrorController, SampleFatalException,
 SampleGeneralException 수정
* 불필요해진 FatalException은 제거

sgo8308 committed:
refactor(common): BaseException의 logLevel 변수에 따라 log level을 다르게 찍도록 변경(#9)

sgo8308 committed:
refactor(common): ErrorResult가 LocalDateTime을 그대로 포함하도록 변경 후 그에 따른 수정 (#9)

sgo8308 committed:
refactor(common): BaseException을 처리하는 핸들러의 이름과 처리 타입을 명확하게 변경 (#9)

sgo8308 committed:
feat(common): UnExpectedException 추가, BaseException 생성자 추가 (#9)

UnExpectedException
* BaseException을 상속하여 다른 예외들과 마찬가지로 일관되게 처리될 수 있도록 추가

BaseException
* 예상하지 못한 예외를 원인 예외로 등록할 수 있도록 생성자 추가

sgo8308 committed:
feat(common): GlobalControllerAdvice에 unExpectedHandler() 추가 (#9)

GlobalControllerAdvice
* UnExpectedException은 이 핸들러에서 다른 예외들과 같이 일관되게 처리

SampleErrorController
*  UnExpectedExceptionHandler를 테스트할 수 있는 핸들러 메소드 추가

sgo8308 committed:
refactor(common): 로그레벨 4개로 나누고 중복되는 로그 메세지 변수로 빼기 (#9)

sgo8308 committed:
Merge pull request #9 from f-lab-edu/feature/jiwoo_errorpage_setting

예외 발생시 일관된 예외 형태가 출력될 수 있도록 초기 구성

sgo8308 committed:
Merge branch 'feature/jiwoo_project_setting' into feature/jiwoo_testdb_setting

sgo8308 committed:
Merge pull request #10 from f-lab-edu/feature/jiwoo_testdb_setting

단위 테스트 환경 구축

sgo8308 committed:
refactor(user, sample): user 패키지를 sample 패키지로 변경 후 다시 user 패키지 생성

sgo8308 committed:
refactor: datasource url db_example에서 marketgola로 수정

sgo8308 committed:
feat(build.gradle): jacoco 추가

테스트를 충분히 작성했는지 확인하기 위해 jacoco 추가

Closes #16

sgo8308 committed:
feat(build.gradle): jacoco report 관련 설정에 exclude 추가

커버리지 체크가 필요하지 않다고 판단한 패키지 및 클래스 제외

Closes #16

sgo8308 committed:
feat(.github): github actions를 사용하여 ci 구성

Closes #16

sgo8308 committed:
Merge pull request #17 from f-lab-edu/feature/16

jacoco 적용

sgo8308 committed:
Merge branch 'feature/jiwoo_project_setting' into feature/jiwoo_sign_up

sgo8308 committed:
chore: 샘플 클래스들 삭제

sgo8308 committed:
feat: 회원가입 관련 기능 구현

- 회원 가입 관련 controller, service, repository 구현
- 회원 가입시 인풋 데이터 Validation 로직 구현
- 아이디, 이메일, 핸드폰 번호 중복 체크

Closes #15

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,103 @@
+ */
+@Slf4j
+@RestControllerAdvice
+public class GlobalControllerAdvice {
```
리뷰 받고 싶은 내용에 언급한 내용에 대해서 답변 드립니다. 
`extends ResponseEntityExceptionHandler`를 확장 받아서 라이브러리에서 제공하는 기본적인 exception처리를 사용할 수 있구요. 저희 요구사항에 맞게 response를 만들기 위해서는 해당 메서드를 Override하는 형태로 구현합니다.

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/ApiError.java:
```diff
@@ -0,0 +1,37 @@
+    private final LocalDateTime timestamp;
+    private final HttpStatus status;
+    private final String message;
+    private final String path;
```
FE에서 Api에러의 메시지만 보고 분기처리하긴 쉽지 않을 것 같습니다. 서버에서 error code같은 것을 만들어서 내려주면 FE에서 그 코드를 기반으로 예외 처리를 할 수 있을 겁니다. 

예를 들면 이런 거죠. 
DuplicateUser -> 이게 코드가 될 수 있습니다. 
DUPLICATE_USER -> 이것도 코드가 될 수 있죠. 
하지만, duplicate user because .... -> 만약 이런 메시지들은 FE에서 분기하기 힘듭니다. 

때문에 errorCode 등을 만들어서 관리할 필요가 있을 것 같습니다. 

이건 보통 예외 발생 시 전달해주는 정보입니다.

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,103 @@
+        ApiError apiError = new ApiError(HttpStatus.BAD_REQUEST, "유효하지 않은 입력값입니다.",
+                request.getRequestURI());
+        if (e.getCause() instanceof JsonMappingException) {
+            apiError.addApiSubError(new ValidationFieldError((JsonMappingException) e.getCause()));
```
apiError를 두고, 이를 상속한 구체적인 xxxApiError라는 클래스를 정의하게 되면,? 
여기에 필드로서 ValidationFieldError 등을 받는다면 지금 처리하고 잇는 동작이 공통화 될 수 있을 것입니다. 

위에서 언급한 `ResponseEntityExceptionHandler`를 상속하고 그 메서드 중 `handleHttpMessageNotReadable`를 구현하면 
HttpMessageNotReadableException -> 에서 관련된 정보들을 얻어 볼 수 있을 겁니다.(디버깅해보면서 판단해보면 될듯합니다)

좀 정리하면 ApiError -> ApiSubError로 모든 것을 소화하기 보다? ApiError에 공통 정보를 담고, 이를 상속한 구체적인 예외

예를 들면 `InvalidParameterApiError`라는 클래스를 만들어서 여기에서 필요한 정보들을 담고, 동작들을 수행하도록 책임을 주는 것이 어떨까요

derek-Flab replied:
으흠? 코드를 보다보니, ApiSubError를 상속한 구체 예외가 있군요? 

ApiError -> SubError list -> subError인터페이스를 상속한 구체 예외 종류 파악? 
음 구조 파악이 살짝 복잡한 것 같긴 하네요. 


제 리뷰 포인트는요.
1. 음 우선, 이 경우에만 생각해보면 예외가 발생한 필드들을 조합하여 객체를 만들어서 주는 것도 좋은 방법이지만, 한문장으로 만들어서 메시지로 보여줘도 괜찮지 않을까요?(이건 개인적인 의견)

2. 그리고, ApiSubError 빈 껍데기 인터페이스를 상속하는 구조보다는 
ApiError에 제네릭 T를 활용하는 방안으로도 구조를 고민해보면 어떨까요?

derek-Flab commented on src/main/java/com/flab/marketgola/user/controller/UserJoinController.java:
```diff
@@ -0,0 +1,45 @@
+
+    private final UserService userService;
+
+    @ResponseStatus(HttpStatus.CREATED)
```
어노테이션을 통해서 response status를 정하는 방법을 사용하셨군요. 

아 또 다른 방법도 소개하면요. 
ResponseEntity<>형태로 리턴하면서 HttpStatus와 Header, Body를 구성할 수 있습니다. 
사용성 측면에서 좋았습니다. 
현재 저는 이렇게 모든 response를 ResponseEntity로 구성해서 내려주는 컨벤션을 정하여 개발하고 있습니다. 
참고차 말씀드리고요. 

모든 메서드에 대해서 정상 실행 되었을 때 강제로 지정해주어야 하기 때문에 지정해주어야 하는 곳을 안한다거나 하는 실수를 방지해줄 수 있었습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/user/controller/UserJoinController.java:
```diff
@@ -0,0 +1,45 @@
+@RequiredArgsConstructor
+@RequestMapping("users")
+@RestController
+public class UserJoinController {
+
```
Restful 관점에서 보면 Resource와 아이디 메서드로 기본적인 CRUD를 표현하는데요. 

현재 /users/id-exist와 같은 형태는 추후 API interface에 대한 전체적인 일관성을 가지기 힘들 것 같다는 걱정이 듭니다. 

restful관점에서 정리된 아티클들이 많은데 한번 읽어보고 같이 이야기해보면 좋을 것 같습니다.

derek-Flab replied:
@sgo8308 
resource와 메서드로 API를 설계하고 그 기준에 맞춰 controller를 만드는게 익숙하다보니 낯선 점이 있었습니다.

우선 UserJoinController는 유저가입에만 한정된 컨트롤러로 활용될 가능성이 있구요. 앞으로 발생할 여러 종류의 기능들과 controller를 맵핑하여야 일관성을 만족하며 전체적인 코드가 만들어질 것인데, 경우에 따라서 controller 개수가 너무 많아질 것 같습니다. 

또 이게 도메인 기반으로 기능을 찾아보기 위해 코드를 찾아들어갈 때쉬운 방법은 아닐 것 같습니다.(제가 이런 구조에 익숙하지 않아서 일수도 있지만, 많은 개발자들이 그렇게 느낄 것 같아요)

리소스별로 표현하는 것으로 출발해보면 어떨까요

현재의 resource design은 이렇게 표현할 수 있을 것 같습니다. 

`(POST) /users` => 유저 생성 (가입에 활용)
`(GET) /users/{id}` => 특정 유저 단건 조회
`(GET) /users?userId=abcdf` => 유저 목록 조회(userId조건)
`(GET) /users?email=abc@flab.com` => 유저 목록 조회(userId조건)
`(GET) /users?phoneNumber=01012345678` => 유저 목록 조회(userId조건)

이렇게 되면 유저 목록 조회 API하나에 여러 가지 파라미터를 제공해주므로서 여러 개의 API엔드포인트를 관리할 필요는 없어지면서
api interface도 사용측면에서도 단순해질 것 같습니다. 

한번 고민해보면 좋겠습니다. 

`best practice rest api`이런식으로 검색해서 한번 아티클들을 읽어보시도 도움이 될 것 같습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/user/controller/UserJoinController.java:
```diff
@@ -0,0 +1,45 @@
+    private final UserService userService;
+
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping()
```
`()`는 빼도 되겠죠?

derek-Flab commented on src/main/java/com/flab/marketgola/user/domain/Membership.java:
```diff
@@ -0,0 +1,12 @@
+public class Membership {
+
+    private String level;
+    private float pointRate;
```
`float`자료형보다는 더 정밀한 자료형을 쓰는게 낫지 않을까요? 
그리고, 필요하다면 Rate에 대한 정확한 연산 등을 위해서 BigDecimal을 쓰는 것도 좋을 것 같네요.

derek-Flab commented on src/main/java/com/flab/marketgola/user/domain/ShippingAddress.java:
```diff
@@ -0,0 +1,22 @@
+@NoArgsConstructor
+public class ShippingAddress {
+
+    Long id;
```
기본적으로 자바 오브젝트의 필드들은 외부에서 직접 접근하지 않게 하려고 private를 꼭 붙여주는 것이 관례입니다. 특별히 붙여주지 않아야 할 이유가 없다면 private을 붙여주는게 좋겠네요.

derek-Flab commented on src/main/java/com/flab/marketgola/user/dto/UserJoinDto.java:
```diff
@@ -0,0 +1,72 @@
+
+    @NotBlank(message = "아이디를 입력하지 않았습니다.")
+    private String loginId;
+    @Pattern(regexp = "^(?!((?:[A-Za-z]+)|(?:[~!@#$%^&*()_+=]+)|(?:[0-9]+))$)[A-Za-z\\d~!@#$%^&*()_+=]{10,64}$", message = "10자 이상 64자 이하의 영문/숫자/특수문자(공백 제외)만 허용하며, 2개 이상 조합해주세요.")
```
오호~이렇게 정규표현식 패턴을 적용해서 validation check를 하셨군요! 좋습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/user/service/UserService.java:
```diff
@@ -0,0 +1,77 @@
+    private Exception identifyError(DuplicateKeyException e) {
+        String errorField = extractErrorField(e);
+        if (errorField.equals("login_id")) {
+            return new DuplicatedLoginIdExcepiton(LogLevel.DEBUG);
```
DuplicatedLoginIdExcepiton에 디폴트로 `LogLevel.DEBUG`를 지정해두면 사용할 때마다 loglevel을 보낼 필요는 없을 것 같습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/user/service/UserService.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private String extractErrorField(DuplicateKeyException e) {
+        Pattern pattern = Pattern.compile("(?<=user.)[a-z_]+(?=_UNIQUE)");
```
음 이렇게 메시지를 발라내서 예외를 만들어주는 것이 유지보수도 힘들고 해서 불가피한 상황(사내 레거시 시스템의 문제)을 제외하고는 이런 구현방식은 피하는게 좋을 것 같습니다. 버전이 변경될 때 exception message가 달라진다면? 또 다른 duplicated 컬럼체크 스펙이 추가될 때마다? 로직이 수정되어야 하는 것에 대한 부담이 있을 것 같습니다. 

1. 제 생각에는 API로 중복 체크를 하더라도, 생성 시점에도 중복체크를 해주고, 간발에 차이로 중복 요청이 들어가는 경우에는 ? 발생하는 예외를 exception handle등에서 duplicated user정도로만 퉁쳐서 처리해도 문제 없을 것 같습니다!

2. 현재 구조 많은 수고와 노력을 해주셨지만, 유지보수하기 힘든 이유로 좀 더 쉽게 구조를 가져가면 좋겠습니다. 
만약 정말 필요하다면, factory method 등을 이용해서 예외를 던지면 DuplicatedxxxException을 만들어내는 구조를 가도록 하면 좋을 것 같아요. factory method등 한번 찾아봐서 아이디어를 얻어보면 좋을 것 같습니다. 

제 의견으로는 1번 방향이 좋을 것 같습니다.

derek-Flab commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,6 @@
+# Profile
+spring:
+  profiles:
+    active: local
```
엇 이거 동작하나요? 
spring boot 구버전 아니고서는 동작 안하는 걸로 알고 있는데요!
다시 한번 확인해보면 좋을 것 같아요. 

그리고 각 profile에 profiles를 전부 명시해주어야 할 것 같습니다.

sgo8308 replied:
확인해보니 deprecated된 것이 맞는 것 같습니다. 그런데 공식 문서에서 확인해보면 여전히 [이 방식](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.profiles:~:text=spring%3A%0A%20%20profiles%3A%0A%20%20%20%20active%3A%20%22dev%2Chsqldb%22)으로 설명하고 있는데요. 최신 방법을 볼 수 있는 문서가 여기 말고 따로 있을까요?

그리고 각 profile에 profiles를 명시해주지 않아도 메인 레파지토리에서는 application.yml에서 프로파일을 정하면 그 다음 순서로 그에 맞는 appication-local등을 찾게 되고, 테스트 레파지토리에서는 프로파일에 상관 없이 각 테스트 클래스 위에 @ActiveProfiles에 맞춰서 application.yml을 선택하더라구요. @ActiveProfiles("unit")이면 spring.active.profiles가 test여도 application-unit.yml을 선택하는거죠.

그래도 각 profile에 profiles를 전부 명시해야할 필요가 있을까요?

forest-lim-dev replied:
@sgo8308 
1. 링크가 정확히 안걸리는거 같은데, https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config.files.profile-specific 이 부분을 말씀하시는건가요? 
저도 이 글이 살짝 해석하기가 난해하긴 한데, 내용을 보자면, 

spring.profiles.active and spring.profiles.default can only be used in non-profile specific documents. This means they cannot be included in [profile specific files](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config.files.profile-specific) or [documents activated](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config.files.activation-properties) by spring.config.activate.on-profile.

==> non-profile 특정 도큐먼트에서만 사용할 수 있다고 하니, 저희가 생각하는 프로파일 설정 파일에 사용할 수 있는 건 아닌거 같습니다. 

https://spring.io/blog/2020/08/14/config-file-processing-in-spring-boot-2-4
이 글을 참고 해보세요!

2. `그래도 각 profile에 profiles를 전부 명시해야할 필요가 있을까요?`
=> 저희는 최종 jar파일을 만들고 이를 실행시킬 계획입니다.
local에 배포할 소스에는 application.yml에 local을 지정하고, beta에 배포할 소스에는 또 해당 소스의 application.yml에 beta라고 수정하고 다시 배포를 할수는 없을겁니다. 저희는 빌드된 하나의 결과물이 있고(예: marketgola-api-0.0.1.jar) 여기에 profile옵션을 주어 실행시켜야 하는거죠. (`java -jar maket-gola-api-0.0.1.jar -Dspring.profiles.active=prod`이런식으로요) 

그렇기 때문에 application.yml에 profile을 바꾸어가면서 사용할 순 없을 것 같고요. 
activeProfiles라는 어노테이션을 설정하면 그 것이 환경변수가 되는 것입니다. 그러므로 test 프로파일이어도? unit이 되는 것이죠!

3. 스프링 부트의 환경설정 변수 관리에 대해서 한번 다음의 내용을 찾아보면 좋을 것 같네요. 
(1) spring-boot에 active profile환경 변수 설정 주입을 어떻게 할까?
(2) 주입된 profile 변수를 기반으로 스프링 부트는 환경설정 (변수) 파일을 어떻게 읽어갈까?
(3) application.yml에 local, dev, prod 환경 변수를 모두 담아 하나의 파일로 관리하려면 어떻게 해야할까?
(4) application.yml, application-dev.yml, application-prod.yml 처럼 분리된 파일이 있을 때 스프링 부트는 프로파일에 따라 어떤 실행 파일을 읽을지 어떻게 판별할까?((2)번과 같은 질문)

4. 
`profile에 profiles를 전부 명시해주어야`
이렇게 말씀 드린 의도는 
application.yml, application-local.yml, application-dev.yml, application-prod.yml 구분되어 있을 때 
active profiles에 따라 알맞은 설정 파일을 읽어가기 때문에 문제는 없습니다. 
(application.yml은 모든 profile이 읽어감, local profile이면 application-local.yml의 변수들을 읽어감 등)
(1) `spring.config.activate.on-profile={프로파일}`을 각각의 파일(application-{프로파일}.yml)에 명시하면 좋겠다고 한건 명시해두면 파일 자체를 판독하는데 좋겠다는 원론적인 의미였고, 반드시 명시해줄 필요는 없습니다. 
(2) 다만, application.yml에 `---`을 구분으로 하여 각각 profile을 명시하는데, 이 때는 반드시 `spring.config.activate.on-profile={프로파일}`변수를 각 프로파일(환경)별로 명시해주어야 합니다.

forest-lim-dev replied:
쫌 커뮤니케이션이 원활하지 않은 부분은 멘토링 시간에 보충해보도록 하시죠!

sgo8308 replied:
1.  spring-boot에 active profile환경 변수 설정 주입을 어떻게 할까?
- 프로그램 안에서 세팅하기
@Autowired
private ConfigurableEnvironment env;
...
env.setActiveProfiles("someProfile");

- JVM 시작할 때 주입하기 -Dspring.profiles.active=dev
- application.properties 또는 application.yml에 선언하기
spring:
  profiles:
    active: dev

=> JVM 시작할 때 주입하는게 말씀하신 것처럼 빌드된 하나의 결과물을 가지고 작업할 때 좋은 방식 같습니다.

2. 주입된 profile 변수를 기반으로 스프링 부트는 환경설정 (변수) 파일을 어떻게 읽어갈까?

주입된 profile변수에 상관없이 처음 application.yml이나 application.properties 파일을 찾아 먼저 읽는다. 그 후 application-local.yml 과 같은 네이밍의 파일이 있을 때 application-뒤를 살펴보고 주입된 profile에 맞는 프로퍼티 파일을 읽는다. 예를 들어 주입된 profile 변수가 local이라면 application.yml을 먼저 읽고 그 다음 application-local.yml을 읽는다.

혹시 이게 아니라 더 깊이 내부적으로 스프링 부트가 어떤 식을 파일을 읽는지 찾아보라는 질문일까요?

3. application.yml에 local, dev, prod 환경 변수를 모두 담아 하나의 파일로 관리하려면 어떻게 해야할까?

--- 로 각 프로퍼티 파일을 구분한다.
각 프로퍼티 파일에서는 spring.config.activate.on-profile을 이용해서 어떤 프로파일에서 활성화되는지 명시한다. 

=> 한 파일에 넣는 방식도 좋지만 저는 따로 따로 나누는 것이 더 보기 편한 것 같아서 따로 따로 나누었습니다.

derek-Flab replied:
`이게 아니라 더 깊이 내부적으로 스프링 부트가 어떤 식을 파일을 읽는지 찾아보라는 질문일까요?`
아닙니다. 어떻게 작동하는지 정도만 레퍼런스 문서를 보고 학습하는 것을 요청드린 겁니다. 
👍

derek-Flab reviewed (commented):
아 아직 전부 리뷰를 하진 못했는데요. 봐야할게 좀 되네요~ㅎㅎ아직 완료되지 않았으니 리뷰는 내일중으로 계속 진행해볼게요.

derek-Flab commented on src/main/java/com/flab/marketgola/user/util/PasswordEncrypter.java:
```diff
@@ -0,0 +1,44 @@
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+public class PasswordEncrypter {
```
util 클래스는 명시적으로 `PasswordEncryptionUtil`로 네이밍해 이것이 spring bean이 아니라, util클래스라는 것을 보여주는 편이 어떨까요?

derek-Flab commented on src/main/java/com/flab/marketgola/user/util/PasswordEncrypter.java:
```diff
@@ -0,0 +1,44 @@
+
+    public static String encrypt(String password)
+            throws NoSuchAlgorithmException, InvalidKeySpecException {
+        int iterations = 1000;
```
1000과 같은 매직넘버, 리터럴의 경우 private static final 상수로 빼서 이름을 지어주는게 가독성 측면에서 좋은 것 같습니다!

같은 클래스의 다른 코드들에서도 숫자들 그리고 ""리터럴 문자들을 상수화해서. 이게 어떤 의미의 상수다? 라는 것을 명확히 표현해주면 좀 더 코드 가독성이 올라갈 것 같네요.

derek-Flab commented on src/test/java/com/flab/marketgola/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,234 @@
+
+@ActiveProfiles("unit")
+@WebMvcTest(UserJoinController.class)
+class UserControllerTest {
```
일관성 있게 네이밍하려면 `UserJoinControllerTest`가 맞지 않을까요?

derek-Flab commented on src/test/java/com/flab/marketgola/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    @DisplayName("로그인 id가 형식에 어긋날 경우 가입에 실패한다.")
+    @ParameterizedTest
```
오~이런 테스트 기능들도 쓰셨군요! 👍

derek-Flab commented:
단위 테스트까지 꼼꼼히 좋네요! 
수고하셨습니다. 리뷰 드린 내용은 반드시 이렇게 해야한다기 보다는 일반적인 관점을 설명드린 부분도 있어 한번 검토후에 개선이 될만한 곳들은 개선을 해보면 좋겠네요.

derek-Flab commented on src/main/java/com/flab/marketgola/user/util/PasswordEncrypter.java:
```diff
@@ -0,0 +1,44 @@
+
+        // 추후 하드웨어 성능 향상으로 인해 iteration이 더 커지더라도
+        // 이전 iteration으로 암호화된 비밀번호를 알아내기 위하여 iteration과 digest를 같이 보관
+        return iterations + ":" + toHex(salt) + ":" + toHex(digest);
```
String.format("%s:%d", iterations, , )
이렇게 쓰면 수정하거나 가독성이 조금이라도 더 좋아질 수 있을 것 같네요.

sgo8308 committed:
refactor(common): GlobalControllerAdvice 관련 리뷰 반영

Closes #15

sgo8308 committed:
refactor(user): 중복체크 관련 리뷰 반영

Closes #15

sgo8308 committed:
refactor : float 관련 리뷰, 매직 리터럴 관련 리뷰 반영

sgo8308 committed:
chore: profile 관련 리뷰 반영

- application.yml 프로파일 결정 속성 삭제
jvm 시작시 프로퍼티 주입하는 방식으로 변경
공통된 속성만 정의할 수 있도록 # Common 주석만 남겨 놓음

- 프로퍼티 파일들에 명시적으로 activate on-profile 속성 추가

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/ApiError.java:
```diff
@@ -0,0 +1,26 @@
+    public ApiError(HttpStatus status, Exception exception, WebRequest request) {
+        timestamp = LocalDateTime.now();
+        this.status = status;
+        this.errorCode = exception.getClass().getSimpleName();
```
예를들어, UserNotFoundException보다는 UserNotFound만 에러 코드로 활용하는게 어떨까요?
에러 코드를 보는 입장에서는 Exception이라는 단어를 제외하고 처리하는게 직관적이고, 클래스명을 유추하는 일이 없을 것 같습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,115 @@
+            log.warn(logMessage, e);
+        } else if (logLevel == LogLevel.INFO) {
+            log.info(logMessage, e);
+        } else {
```
LogLevel중 아예 서버에서 로깅하지 않는 레벨도 관리하면 완성도가 높을 것 같습니다.
뭐 예를 들어, 개발환경에서만 찍히겠지만, "사용측에서 API request정보를 잘못입력한 경우 서버에서는 특별히 로깅하지 않겠다."라는 니즈가 있을 수도 있으니까요!

derek-Flab commented on src/main/java/com/flab/marketgola/user/controller/UserController.java:
```diff
@@ -0,0 +1,40 @@
+    private final UserService userService;
+
+    @PostMapping
+    public ResponseEntity<Void> joinUser(
```
1. 유저를 생성하는 경우에 response body는 userDto를 내려주는게 좋을 것 같습니다. (가입 후 OK면 다시 조회하기 때문에? 네트워크 비용을 아끼려는 의도입니다.)
2. 다른  이야기로 저는 메서드 명만으로 controller, service, repository를 구분하기 위해 컨벤션을 정해놓는 타입입니다. 의외로 가독성에 도움이되서 추천드립니다. 

controller, service는 조회(get), 생성(create), 수정(modify), 삭제(remove)
repository는 find, insert, update, delete -> SQL의 예약어를 그대로 사용하는? 또는 JPA와 같은 ORM의 예약어를 사용하는 형태?

이렇게 컨벤션을 정하게 되면, 메서드명 만으로 충분히 어떤 레이어인지 판단이 되서 가독성측면에서 좋았던 것 같습니다. 

위의 단어를 그대로 사용하는 것이 아니라, 위의 형태처럼 service <-> repository(mapper)를 구분하면 좋겠다는 의견입니다.

derek-Flab commented on src/main/java/com/flab/marketgola/user/exception/DuplicatedLoginIdExcepiton.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.boot.logging.LogLevel;
+import org.springframework.http.HttpStatus;
+
+public class DuplicatedLoginIdExcepiton extends BaseException {
```
오타 제보입니다.
DuplicatedLoginIdExcepiton -> DuplicatedLoginIdException

derek-Flab commented on src/main/java/com/flab/marketgola/user/exception/DuplicatedLoginIdExcepiton.java:
```diff
@@ -0,0 +1,14 @@
+
+public class DuplicatedLoginIdExcepiton extends BaseException {
+
+    public DuplicatedLoginIdExcepiton() {
```
만약 LogLevel.DEBUG 이외에 LogLevel이 필요한 경우 파라미터로 받는 생성자도 고려하고 계신거죠 ?ㅎㅎ

sgo8308 replied:
네 ! 필요한 경우가 생기면 그 때 추가하려구요.

derek-Flab reviewed (approved):
위의 추가적인 리뷰 사항만 수정되면 더 리뷰드릴 내용은 없을 것 같습니다. 
검토해주세요. 
수고하셨습니다.

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/InvalidArgumentApiError.java:
```diff
@@ -0,0 +1,29 @@
+@Setter
+public class InvalidArgumentApiError extends ApiError {
+
+    private List validationErrors = new ArrayList();
```
Generic 타입 명시하는게 좋을 것 같습니다. 가능하면 타입은 반드시 명시하는 것으로!

sgo8308 replied:
이 리스트에는 ValidationFieldError말고도 ValidationGlobalError가 들어갈 수 있는데요 그럼 둘의 공통 부모를 다시 정의해서 제네릭 타입으로 선언해주어야 할까요? 
T extends ValidationError 이렇게요

빈 껍데기 인터페이스를 만들고 싶지 않아서 그냥 제네릭 없이 구현했는데 고민이 되네요 !

derek-Flab commented on src/main/java/com/flab/marketgola/common/exception/InvalidFormatApiError.java:
```diff
@@ -0,0 +1,24 @@
+        super(httpStatus, exception, request);
+
+        String field = exception.getPath().get(0).getFieldName();
+        String message = "잘못된 타입의 입력값입니다.";
```
매직 넘버, 문자열 리터럴은 private static final 상수로 빼면 더 좋을 것 같습니다!

forest-lim-dev commented:
```
GlobalControllerAdvice 관련
ApiSubError는 삭제하고 각 에러마다 클래스를 만들어주었는데요.(InvalidArgumentApiError, InvalidFormatApiError)
이렇게 하니 에러 메시지에 일관성이 좀 깨지는 것 같습니다.
InvalidArgumentApiError은 필드로 validationErrors를 갖고 있고, InvalidFormatApiError은 validationError를 갖고 있습니다. 추가적으로 다른 에러가 생길 때 마다 이 추가 서브 에러들의 이름이 바뀌면 일관성이 없지 않을까요? 기존에는 모두 subApiErrors로 통일했었습니다.
멘토님께서 얘기해주신 하나의 string으로 일관되게 반환할 수 있지만 그럴 경우 InvalidArgumentApiError에서 여러 errorfield와 rejected value가 하나의 문장 안에 들어 있어서 프론트 엔드 쪽에서 파싱하기가 힘들 것 같습니다.
```
==> 
제 생각에는 ApiError에 

`protected T detail;` 필드를 추가하고 확장 InvalidArgumentApiError같은 곳에서 이 필드를 채워주면 되지 않을까 싶습니다. 
이렇게 되면 detail이라는 필드에 예외 객체마다 조금 다른 타입의 내용을 담고 잇을 수 있지만, 필드 이름이 바뀌진 않을 것 같고요. 

```
public class ApiError<T> {
....
    private final String errorCode;
    private final String path;
    protected T detail;
....
}

public class InvalidArgumentApiError extends ApiError<List<ValidationFieldError>> {

    private List<ValidationFieldError> validationErrors = new ArrayList<>();
....
```

이런식으로요? 

```
중복 체크 관련
중복체크를 /users?loginId= ~ 이런 api로 처리하게끔 구현하고 response로는 id만 나가도록 구현했는데요. 이 /users라는 api가 중복체크 외에도 실제로 여러 유저들을 특정한 조건으로 찾고 싶을 때도 쓸 수 있을 것 같고 여러 용도로 쓰일 수 있을 것 같다는 생각이 들었습니다.
그럴 경우 각 경우마다 response가 달라질 수 있을 것 같다는 생각이 들어 클라이언트 쪽에서 원하는 field만 쿼리 스트링에 담아서 보내는 식으로 구현해야 할까 생각이 들었습니다. response는 그에 따라 동적으로 달라지게 되구요. 구현하게 되면 API는 /users?loginId='abc123'&name='지우'&fields=id,loginId,name,address 이런 식으로 사용될 것 같습니다. 하지만 아직 뚜렷하게 이 API를 필요로 하는 다른 기능이 생각이 나지 않아 오버 엔지니어링일 것 같아 지금과 같이 구현하였습니다. 멘토님은 어떻게 생각하시나요?
```
==> 
아이고 이 부분을 못봤네요 ㅎㅎ 응답에 id만 있었군요ㅠ 
response는 고정적인 UserResponse가 나가야됩니다. 
리소스를 요청했으니 리소스를 응답해주는 것이 맞는 것 같습니다. 
요청마다 응답이 달라질 필요는 없을 것 같습니다? 
만약 이렇게 응답이 달라지는 것이라면? API를 분리해야할 필요가 있을 것 같습니다. 
기본적인 사상은 users리소스를 요청하는 API인터페이스니, 전부 userResponse?(userDTO)를 리턴해주는 것이 맞습니다!

sgo8308 committed:
feat(common): ApiError의 ErrorCode를 파싱하는 메소드 추가

불필요한 Excepion이라는 부분을 제외하는 것 이 직관적이고 클라이언트 쪽에서 클래스명을 유추할 수  없음

sgo8308 committed:
feat(common): GlobalControllerAdvice에서 아예 로그를 찍지 않는 옵션도 추가

특별히 로그를 찍을 필요가 없는 상황도 있을 수 있기 때문

sgo8308 committed:
feat(user): 유저 가입과 유저를 찾는 요청시 UserResponseDto를 반환하도록 변환

가입할 경우 회원가입 정보를 표시해야 하느라 가입된 정보가 필요하다.
유저를 찾을 경우 유저의 최소한 기본적인 정보가 필요하다.

sgo8308 committed:
refactor(user): controller, service와 mapper 클래스의 메소드 네이밍을 다르게 가져감

메서드만으로 어떤 레이어에 메소드인지 파악 가능하여 더 가독성이 좋아진다

sgo8308 committed:
style(user): DuplicatedLoginIdExcepiton 클래스명 오타 수정

sgo8308 committed:
refactor(common): ApiError에 detail 필드 추가

ApiError를 상속하는 다른 클래스들이 재정의하게 하여 일관된 에러 메시지를 응답할 수 있게 하기 위하여
